lib/web/service-worker/worker.js

```diff
@@ -0,0 +1,106 @@
+import * as fs from 'node:fs'
+import * as vm from 'node:vm'
+import { workerData, parentPort as maybeParentPort } from 'node:worker_threads'
+import { ServiceWorkerGlobalScope } from './service-worker-global-scope.js'
+import { ExtendableEvent, kPendingPromises } from './extendable-event.js'
+import { InstallEvent } from './install-event.js'
+import { FetchEvent, kResponsePromise } from './fetch-event.js'
+
+const parentPort = maybeParentPort
+
+if (!parentPort) {
+  throw new Error('Failed to run worker: missing parent process')
+}
+
+/** @type {import('./ServiceWorkerContainer.js').WorkerData} */
+const parentData = workerData
+
+// Create the Service Worker's global scope object (`self`).
+const globalScope = new ServiceWorkerGlobalScope(parentData)
+
+process.once('uncaughtException', (error) => {
+  console.error(error)
+  globalScope.serviceWorker.dispatchEvent(new Event('error'))
+})
+
+const content = fs.readFileSync(parentData.scriptUrl, 'utf8')
+
+parentPort.postMessage({
+  type: 'worker/statechange',
+  state: 'parsed'
+})
+
+// Run the worker script within the controller global scope.
+const script = new vm.Script(content)
+
+script.runInNewContext({
```

Comment:
This is the heart of the implementation: running the actual worker script in a VM _inside_ a `worker.js` context. See `service-worker-container.js` for more details on how `worker.js` is spawned.